File: src/est/tsls.r
# setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
#########################################################################
# Dependencies

#########################################################################
# 2-Stage Least Squares

tsls <- function(y, X, Z, ...) {
  n <- nrow(Z)
  Z <- cbind(ones(n), Z)
  X <- cbind(ones(n), X); Xt <- t(X); XtX <- Xt %*% X
  num <- Xt %*% P(Z, y) 
  denom <- Xt %*% P(Z, X)
  delta.hat <- solve(denom, num) %>%
    as.numeric
  # theta.hat <- (num/denom) %>% as.numeric
  
  sigma.hat2 <- sum((y - X %*% delta.hat)^2) / (n-1)
  sigma.hat <- sqrt(sigma.hat2)
  SE_delta.hat <- (sigma.hat2 * solve(denom)) %>% diag %>% sqrt
  
  list(theta.hat = delta.hat[2], sigma.hat = sigma.hat, 
       SE_theta.hat = SE_delta.hat[2], ...)
}

File: src/est/tsls.jl
module TSLS

struct TSLSFit{T<:Real}
  δ̂::Vector{T}
  σ̂::T
  Λ̂::Matrix{T}
end


function tsls(y, X, Z)
  n, K = size(Z)
  ZZ = Z' * Z
  Xt = X'
  num = Xt * Z * (ZZ \ Z'*y)
  denom = Xt * Z * (ZZ \ Z'*X)
  δ̂ = denom \ num

  σ̂2 = sum((y - X * δ̂).^2) / (n-1)
  σ̂ = sqrt(σ̂)
  Λ̂ = σ̂ * inv(denom)

  return TSLSFit(δ̂, σ̂, Λ̂)
end

end
